File: src/main/java/org/micoli/php/symfony/cliDumpParser/actions/AbstractParseCliDumperAction.java
package org.micoli.php.symfony.cliDumpParser.actions;

import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;

public abstract class AbstractParseCliDumperAction extends AnAction {

    @Override
    @NotNull
    public ActionUpdateThread getActionUpdateThread() {
        return ActionUpdateThread.BGT;
    }

    @Override
    public void actionPerformed(@NotNull AnActionEvent e) {
        String selection = getCurrentSelection(e);
        if (selection == null) {
            return;
        }
        parseCliDumperOutput(selection, e.getProject());
    }

    @Override
    public void update(@NotNull AnActionEvent e) {
        boolean isVisible = getCurrentSelection(e) != null;
        e.getPresentation().setVisible(isVisible);
        e.getPresentation().setEnabled(isVisible);
        setPresentationText(e.getPresentation(), isVisible);
    }

    private String getCurrentSelection(@NotNull AnActionEvent e) {
        Project project = e.getProject();
        Editor editor = e.getData(CommonDataKeys.EDITOR);

        if (editor == null || project == null) {
            return null;
        }
        String selectedText = editor.getSelectionModel().getSelectedText();
        if (selectedText == null || selectedText.trim().isEmpty()) {
            return null;
        }

        return selectedText;
    }

    abstract protected void parseCliDumperOutput(String text, Project project);

    abstract protected void setPresentationText(@NotNull Presentation presentation, boolean isVisible);
}

File: src/main/kotlin/org/micoli/php/symfony/cliDumpParser/actions/AbstractParseCliDumperAction.kt
package org.micoli.php.symfony.cliDumpParser.actions

import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.actionSystem.Presentation
import com.intellij.openapi.project.Project

abstract class AbstractParseCliDumperAction : AnAction() {
    override fun getActionUpdateThread(): ActionUpdateThread {
        return ActionUpdateThread.BGT
    }

    override fun actionPerformed(e: AnActionEvent) {
        val selection = getCurrentSelection(e) ?: return
        val project = e.project ?: return
        parseCliDumperOutput(selection, project)
    }

    override fun update(e: AnActionEvent) {
        val isVisible = getCurrentSelection(e) != null
        e.presentation.isVisible = isVisible
        e.presentation.isEnabled = isVisible
        setPresentationText(e.presentation, isVisible)
    }

    private fun getCurrentSelection(e: AnActionEvent): String? {
        val project = e.project
        val editor = e.getData(CommonDataKeys.EDITOR)

        if (editor == null || project == null) {
            return null
        }
        val selectedText = editor.selectionModel.selectedText
        if (selectedText == null || selectedText.trim { it <= ' ' }.isEmpty()) {
            return null
        }

        return selectedText
    }

    protected abstract fun parseCliDumperOutput(text: String, project: Project)

    protected abstract fun setPresentationText(presentation: Presentation, isVisible: Boolean)
}
